server/apiserver/keepalive.go

```diff
@@ -0,0 +1,94 @@
+package apiserver
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/felixge/httpsnoop"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/argoproj/argo-workflows/v3/util/ticker"
+)
+
+type tickerFactoryFn func(time.Duration) ticker.Ticker
+
+func serverSentEventKeepaliveMiddleware(next http.Handler, keepaliveInterval time.Duration) http.HandlerFunc {
+	return serverSentEventKeepaliveMiddlewareAux(next, keepaliveInterval, nil, func(d time.Duration) ticker.Ticker {
+		return ticker.NewTicker(d)
+	})
+}
+
+func serverSentEventKeepaliveMiddlewareAux(next http.Handler, keepaliveInterval time.Duration, wg *sync.WaitGroup, tickerFactory tickerFactoryFn) http.HandlerFunc {
```

Comment:
can we update StreamForwarder to have this feature instead?

https://github.com/argoproj/pkg/blob/master/grpc/http/forwarders.go#L136

Reply:
I'm open to that and it'd actually simplify the code a bit (no redundant `Content-Type` header checks needed).  The reason that I didn't in this initial implementation was that the stream forwarder is located in a shared package, and I was trying to limit the blast radius of this change to get some initial testing before potentially moving it there.  Happy to do that instead of it's preferable.

Reply:
That said, one potentially large benefit of doing it in the StreamForwarder is I would no longer need `httpsnoop` since I know exactly when data is being written to the response, so I would no longer need to wrap the ResponseWriter at all.

Reply:
https://github.com/argoproj/pkg/pull/35